src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.cs

```diff
@@ -498,50 +498,70 @@ public void OnCompleted(Func<object, Task> callback, object state)
             }
         }
 
-        protected async Task FireOnStarting()
+        protected Task FireOnStarting()
         {
-            Stack<KeyValuePair<Func<object, Task>, object>> onStarting = null;
+            Stack<KeyValuePair<Func<object, Task>, object>> onStarting;
             lock (_onStartingSync)
             {
                 onStarting = _onStarting;
                 _onStarting = null;
             }
-            if (onStarting != null)
+
+            if (onStarting == null)
             {
-                try
-                {
-                    foreach (var entry in onStarting)
-                    {
-                        await entry.Key.Invoke(entry.Value);
-                    }
-                }
-                catch (Exception ex)
+                return TaskCache.CompletedTask;
+            }
+            else
+            {
+                return FireOnStartingAwaited(onStarting);
+            }
+        }
+
+        private async Task FireOnStartingAwaited(Stack<KeyValuePair<Func<object, Task>, object>> onStarting)
+        {
+            try
+            {
+                foreach (var entry in onStarting)
                 {
-                    ReportApplicationError(ex);
+                    await entry.Key.Invoke(entry.Value);
                 }
             }
+            catch (Exception ex)
+            {
+                ReportApplicationError(ex);
+            }
         }
 
-        protected async Task FireOnCompleted()
+        protected Task FireOnCompleted()
         {
-            Stack<KeyValuePair<Func<object, Task>, object>> onCompleted = null;
+            Stack<KeyValuePair<Func<object, Task>, object>> onCompleted;
             lock (_onCompletedSync)
             {
                 onCompleted = _onCompleted;
                 _onCompleted = null;
             }
-            if (onCompleted != null)
+
+            if (onCompleted == null)
+            {
+                return TaskCache.CompletedTask;
+            }
+            else
+            {
+                return FireOnCompletedAwaited(onCompleted);
+            }
+        }
+
+        private async Task FireOnCompletedAwaited(Stack<KeyValuePair<Func<object, Task>, object>> onCompleted)
+        {
+            foreach (var entry in onCompleted)
             {
-                foreach (var entry in onCompleted)
+                try
```

Comment:
Why is the try/catch inside the foreach in this method and outside the foreach in `FireOnStartingAwaited`?